backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchProjectionFactoryImpl.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.search.query.impl;
```

Comment:
I did move the class but I'm not sure to understand the problems you envisioned as simply moving the class worked?

Reply:
Just moving the class introduces a cyclic dependency, as mentioned in item 1 above; you now have:

* `org.hibernate.search.backend.elasticsearch.search.query.impl.ElasticsearchSearchTargetContext`
* depends on `org.hibernate.search.backend.elasticsearch.search.projection.impl.ElasticsearchSearchProjectionFactoryImpl`
* depends on `org.hibernate.search.backend.elasticsearch.search.query.impl.SearchBackendContext` (same package as `ElasticsearchSearchTargetContext`)

To remove the cyclic dependency, you would need to declare a `SearchProjectionBackendContext` interface in the `org.hibernate.search.backend.elasticsearch.search.projection.impl` package, make `SearchBackendContext` create one such instance in its constructor, and pass the "backend projection context" only to the `ElasticsearchSearchProjectionFactoryImpl` constructor. So that the dependency is "query => projection", like everywhere else, and not the opposite. But really, I think removing the optimization altogether would be simpler.

Reply:
Just moving the class introduces a cyclic dependency, as mentioned in item 1 above; you now have:

* `org.hibernate.search.backend.elasticsearch.search.query.impl.ElasticsearchSearchTargetContext`
* depends on `org.hibernate.search.backend.elasticsearch.search.projection.impl.ElasticsearchSearchProjectionFactoryImpl`
* depends on `org.hibernate.search.backend.elasticsearch.search.query.impl.SearchBackendContext` (same package as `ElasticsearchSearchTargetContext`)

To remove the cyclic dependency, you would need to do as explained above:

> declare a SearchProjectionBackendContext interface in the projection (`org.hibernate.search.backend.elasticsearch.search.projection.impl`) package, make SearchBackendContext create one such instance in its constructor, and pass the "backend projection context" only to the ElasticsearchSearchProjectionFactoryImpl constructor.

Reply:
Just moving the class introduces a cyclic dependency, as mentioned in item 1 above; you now have:

* `org.hibernate.search.backend.elasticsearch.search.query.impl.ElasticsearchSearchTargetContext`
* depends on `org.hibernate.search.backend.elasticsearch.search.projection.impl.ElasticsearchSearchProjectionFactoryImpl`
* depends on `org.hibernate.search.backend.elasticsearch.search.query.impl.SearchBackendContext` (same package as `ElasticsearchSearchTargetContext`)

To remove the cyclic dependency, you would need to do as explained above:

> declare a SearchProjectionBackendContext interface in the projection package (`org.hibernate.search.backend.elasticsearch.search.projection.impl`), make SearchBackendContext create one such instance in its constructor, and pass the "backend projection context" only to the ElasticsearchSearchProjectionFactoryImpl constructor.